### DISCUSSION TITLE
Rows can probably be made an IntoIterator

### DISCUSSION THREAD

**Discussion by althonos** [rusqlite/rusqlite#1102] [Questions] [answered]
Hi there @jgallagher ! Very neat library, I'm glad to have access to a simple and feature-rich Rust API to manage SQLite3 databases, this is very much appreciated.

~I noticed how you mentioned that `Rows` cannot be made an iterator, but actually I think it can: since you only ever send a reference to the internal row once, so you could actually send that row, using `Option.take`.~

~I tried the following code within `row.rs` and it compiled without trouble, so it might be worth looking into that:~
```rust
impl<'stmt> Iterator for Rows<'stmt> {
    type Item = Result<Row<'stmt>>;
    fn next(&mut self) -> Option<Self::Item> {
        match self.advance() {
            Err(e) => Some(Err(e)),
            Ok(_) => self.row.take().map(Ok),
        }
    }
}
```

~Taking ownership of `Row` is not a big deal anyway, since it only wraps a reference, which means once the compiler does its magic there is probably not much difference between `&Row` and `Row` anyway.~

EDIT: Okay, I think I got why this is not possible, but there is probably a way around this without breaking anything, as described below.

I'm happy to send a PR with proper tests if that's something you'd want to include in the code base. Cheers!

**Comment by althonos** [comment]
Or i guess this is related to #150, but then I think there is probably way to at least implement an `IntoIterator` trait using an proxy iterator with a lifetime bound to the lifetime of the rows, e.g.

```rust
impl<'rows, 'stmt> IntoIterator for &'rows Rows<'stmt'> {
   ...
}
```

**Comment by gwenn** [comment]
See [impl<'stmt> FallibleStreamingIterator for Rows<'stmt>](https://docs.rs/rusqlite/0.21.0/rusqlite/struct.Rows.html#implementations) and [fallible_streaming_iterator](https://docs.rs/fallible-streaming-iterator/0.1.9/fallible_streaming_iterator/index.html).

**Comment by thomcc** [comment, marked as answer]
More broadly, an iterator isn't allowed to yield items that are borrowed from the iterator itself. This is a known problem (https://docs.rs/streaming-iterator/0.1.5/streaming_iterator/ has a decent overview of the problem -- the fallible- crate just tweaks the definition for better ergonomics in the fallible case), there isn't really a way to make it compatible.

Note that https://docs.rs/rusqlite/0.21.0/rusqlite/struct.AndThenRows.html is able to be an iterator, if that works for your case.
